Rework build system v2

N-R-K opened the pull request:
As discussed in https://github.com/nsxiv/nsxiv/pull/66#issuecomment-921327511 , this PR takes care of any remaining cleanups on the makefile and cleans `version.h` upon `make clean`.

Also revert back to using `mkdir -> cp -> chmod` since `install` is not posix.

We may want to reconsider reverting `ifeq` as well, see: https://github.com/muennich/sxiv/commit/5155d52ab18c06925046e2f56a87f2765bb515ba

N-R-K commented:
The initial reasoning for having auto library detection was that minimalist distros might benefit from it, but now as it turns out this might've actually harmed those distro users who are using non-gnu system.

I think this should be a good lesson for us moving forward, while certain things might seem "smart" or "convenient" it actually does more harm by bringing unnecessary dependencies.

N-R-K commented:
I don't believe `-include` is standardized either. And using `HAVE_LIBX ?=` instead of `ifeq` can cause potential confusion as described in https://github.com/nsxiv/nsxiv/pull/19#issuecomment-917577494

Overall I think we may just want to revert the auto-detection all together given the fact that running `make HAVE_LIBX=0` isn't hard at all. @TAAPArthur @bakkeby @XPhyro

TAAPArthur commented:
> I don't believe `-include` is standardized either. And using `HAVE_LIBX ?=` instead of `ifeq` can cause potential confusion as described in [#19 (comment)](https://github.com/nsxiv/nsxiv/pull/19#issuecomment-917577494)

Since we auto create the config.mk in the file version, I believe we can drop the `-` since it should always exist/be autocreated. Didn't realize it wasn't portable, but don't think there is anything else non-standard that would warrant reverting the issue.

>I think we may just want to revert the auto-detection all together given the fact that it will make our build more portable and running make HAVE_LIBX=0 isn't hard at all.

I don't want to special case any single package manager/distro, but the reason that this is helpful for KISS Linux, is that the packager generally provides the bare minimum build script/dependencies needed to build the package. So they either have to forcibly list libgif and friends as mandatory or not use them. Something like `make AUTODETECT=1` would fulfill the purpose too if we just didn't want to make it a default.

N-R-K commented:
> is that the packager generally provides the bare minimum build script/dependencies needed to build the package.

AFIAK Kiss packages are just shell/build scripts, so adding `HAVE_LIBX` should be absolutely trivial. Also consider this, `imlib2` already pulls in `giflib` if built with gif support. And `imlib` on the k1ss community repo seems to advise [manually un-commenting it if needed](https://github.com/dylanaraps/community/blob/master/community/imlib2/depends). Is there any reason why `nsxiv` can't be the same?

> Something like make AUTODETECT=1 would fulfill the purpose too if we just didn't want to make it a default.

If we can make it portable without using anything non-standard then, I'm okay with that.

TAAPArthur commented:
>AFIAK Kiss packages are just shell/build scripts, so adding HAVE_LIBX should be absolutely trivial.
> And imlib on the k1ss community repo seems to advise manually un-commenting it if needed.

Yes but requires everyone who wants a different set of depends to fork which is also trivial but extra, unneeded work for what I claim we know the correct behavior. It isn't that I want to choose between portability and convenience but think we can get both.

>If we can make it portable without using anything non-standard then, I'm okay with that.

So we agree to just drop the "-". It is leftover from the draft where there was an optional configure script. Is there any other portability concerns. 

Should we also include `.POSIX` [target](https://pubs.opengroup.org/onlinepubs/009695299/utilities/make.html) to avoid accidentally being non-POSIX

N-R-K commented:
> Is there any other portability concerns.

I couldn't find any info about `ifeq` being standard. And given that https://github.com/muennich/sxiv/commit/5155d52ab18c06925046e2f56a87f2765bb515ba is removing `ifeq`s I am assuming it's non-standard.

TAAPArthur commented:
> > Is there any other portability concerns.
> 
> I couldn't find any info about `ifeq` being standard. And given that [muennich/sxiv@5155d52](https://github.com/muennich/sxiv/commit/5155d52ab18c06925046e2f56a87f2765bb515ba) is removing `ifeq`s I am assuming it's non-standard.

Well I found [mention](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/make.html) that `ifdef` and `include` themselves are an extension so
>Historically, the make utility has been an especially fertile ground for vendor and research organization-specific syntax modifications and extensions. Examples include: 
>...
>Adding C preprocessor-style "include" and "ifdef" constructs (System V, GNU, BSD, and others)

So you are probably right. If we want to be as portable as possible on principle, then I don't object to removing the auto detection. Still think the we should build with minimal dependencies by default.

TAAPArthur commented:
Wait how would optional libs even work them. We need ifs to actually add the libs

N-R-K commented:
> Wait how would optional libs even work then. We need `if`s to actually add the libs.

We don't. We can add `HAVE_LIBX = 1` as it was before. User's would directly edit the makefile, or pass in `make HAVE_LIBX=0`.

So since we can't do any auto-detection without using extensions, are you still opposed to dropping it?

N-R-K commented:
Since you don't oppose it any longer, I've pushed the change.

There's now two more things we need to address,

* `?=`
* and `$(RM)`

Are they standard?

It seems like `?=` was accepted (?) https://austingroupbugs.net/view.php?id=330
I'm not sure about `$(RM)`

eylles commented:
>Adding C preprocessor-style "include" and "ifdef" constructs (System V, GNU, BSD, and others)

wait, doesn't that mean that idef and include are present in bsd?

N-R-K commented:
> wait, doesn't that mean that idef and include are present in bsd?

Yes, as non standard extensions. The `sxiv` makefile previously used to use them, but dropped it in favor of portability and sticking to the standard. See the linked commit in OP.

eylles commented:
Hmm i would preffer to not give up on the auto detection tho, ah btw you can assign variables to the result of shell commands with the syntax:
VAR = \`command\` 
And that is a posix macro, is there any posibility we can use that to still have autodetection but in a more portable way?.
also if gnu make and bsd make have the extensions to use autoreload is not like nsxiv would be compiled in many more systems, would need to check what extensions do other popular make utils have (except windows make) before deciding harshly as iirc there's no make implementation out there that is truly posix make as all have extensions.

N-R-K commented:
> making nsxiv build with minimal dependencies by default would be sufficient.

Taap, I see where you're coming from, and I honestly agree that user's should have to turn optional dependancies on manually. However consider this, if someone is coming over from `sxiv` and he's used to having gif playback by default, simply running `make` will no longer have multiframe gif for him.

One of our project goals is to not break backwards compatibility, I'm not sure if this case qualifies but it's still something we should consider more deeply.

eylles commented:
looking around include is actually POSIX, on the regard of dealing with libraries we could use pkg-config which is supported by all unix like OSes from mac to bsd and the syntax seems easy enough to use.

```
DEP1_CFLAGS = `pkg-config --cflags dep1` -D HAVE_DEP1
DEP1_LDFLAGS = `pkg-config --libs dep1`
```

N-R-K commented:
I don't believe `pkg-config` is posix. `man 1p pkg-config` does not bring up anything.

> looking around include is actually POSIX

Can you link the source?

TAAPArthur commented:
> > making nsxiv build with minimal dependencies by default would be sufficient.
> 
> Taap, I see where you're coming from, and I honestly agree that user's should have to turn optional dependancies on manually. However consider this, if someone is coming over from `sxiv` and he's used to having gif playback by default, simply running `make` will no longer have multiframe gif for him.
> 
> One of our project goals is to not break backwards compatibility, I'm not sure if this case qualifies but it's still something we should consider more deeply.

I suspected the backwards compatibility argument would be made to an attempt to change the defaults which is why I went with auto-detection. But auto-detection was just a means to achieve the goal of not having to explicitly disable the growing list of optional dependencies.

N-R-K commented:
If `include` is posix, then I believe we can have a `make autoconfig` rule, which you said should take care of your use case.

TAAPArthur commented:
> I don't believe `pkg-config` is posix. `man 1p pkg-config` does not bring up anything.

Not POSIX and isn't available on all systems (but probably is on most). Not saying the program itself sucks, but suckless.com promotes a [less sucky alternative to pkg-config](https://suckless.org/other_projects/). I went with using the compiler because we could guarantee it was there. But it really is the same end result either way assuming things are installed in a standard location

TAAPArthur commented:
> If `include` is posix, then I believe we can have a `make autoconfig` rule, which you said should take care of your use case.

Does this imply we are reverting to enabling all optional deps by default?
Also the source I linked stated include was an extension (that basically everyone implemented). And if `include`/`ifeq` are accetable, why not just keep the auto-detection as default as that seems the most friendly behavior.

eylles commented:
so apparently `include` is indeed posix but `-include` isn't

N-R-K commented:
> And if `include`/`ifeq` are accetable, why not just keep the auto-detection as default as that seems the most friendly behavior.

Entirely depends on if `include` is posix or not. `ifeq` i don't think is posix, but we can use `?=` instead to get around that.

N-R-K commented:
Welp, that gets me back to an earlier point about `?=` being a potential source of confusion https://github.com/nsxiv/nsxiv/pull/19#issuecomment-917577494

N-R-K commented:
@TAAPArthur does this work for you? I think the comments should take care of the confusion.

Also I think you misunderstood the manpage, it's talking about options which used to be extensions but are now included in posix. `man 1p make` has `include` defined. So `include` should be posix, but `ifeq` is not defined.


- - -

Finally now the only remaining concern is `$(RM)`.

TAAPArthur commented on Makefile:
```diff
@@ -13,25 +8,26 @@ MANPREFIX = $(PREFIX)/share/man
 # autoreload backend: inotify/nop
 AUTORELOAD = inotify
 
+# include auto-detected libraries
+include lib_autodetect.mk
+
+# default fallback if lib_autodetect.mk was not found
+HAVE_LIBGIF ?= 1
+HAVE_LIBEXIF ?= 1
+
```
If lib_autodetect.mk doesn't exist it will be created at this point. The default vars only matter if they are missing from lib_autodetect.mk and think missing values should air on the safe side (the side more likily to compile) and disable optional libs

N-R-K replied:
Do you have any suggestion on how to better word it then?

TAAPArthur replied:
Well the main suggestion was to switch the default, but for the comment
```
# default fallback if vars not found in lib_autodetect.mk
```
Maybe `HAVE_LIBX` instead of vars?

N-R-K replied:
I think the updated comment should be clear enough.

TAAPArthur commented on Makefile:
```diff
@@ -75,24 +71,28 @@ version.h: Makefile .git/index
-	install -Dt $(DESTDIR)$(PREFIX)/bin nsxiv
+	mkdir -p $(DESTDIR)$(PREFIX)/bin
+	cp nsxiv $(DESTDIR)$(PREFIX)/bin/
+	chmod 755 $(DESTDIR)$(PREFIX)/bin/nsxiv
```
Why use 3 commands instead of 1?

N-R-K replied:
I think I explained that in the OP.

> Also revert back to using mkdir -> cp -> chmod since install is not posix.

TAAPArthur replied:
That you did. My bad. I had thought install was POSIX.

TAAPArthur commented on Makefile:
```diff
@@ -13,25 +8,26 @@ MANPREFIX = $(PREFIX)/share/man
 AUTORELOAD = inotify
 
+# include auto-detected libraries
+include lib_autodetect.mk
```
This will create lib_autodetect even when running make clean. So it would be created in order to be immediately destroyed

N-R-K replied:
Do you have any standard way to avoid this then?

TAAPArthur replied:
Not off hand. Someone pointed it out during the original review so just echoing the concern. I'll put some thought into it later tonight

N-R-K commented:
Related: https://github.com/nsxiv/nsxiv/issues/70#issuecomment-922286267

> I have habit to just run `sudo make clean install` for all suckles software and sxiv. That worked for sxiv but not here.

TAAPArthur commented on Makefile:
```diff
@@ -50,16 +47,17 @@ nsxiv: $(OBJS)
 window.o: icon/data.h
 
-config.mk:
+lib_autodetect.mk:
```
I think `autodetect.mk` is a better name just in case in the feature we want to autodetect inotify

N-R-K commented:
I wasn't against having auto detection as long as we could have it be standard conformant. But after realising that we broke `make clean install` due to this, I've been thinking more about this.

Is auto detection something ***nsxiv*** really needs? Or are we trying to make up for the inadequacies of K1ss package manager?

The more I think about it the more it feels like its the latter case. If the package manger of a source based distro cannot keep track of optional dependencies then (it's a complete failure and) that package manager should be fixed, not the upstream build system.

TAAPArthur commented:
> Is auto detection something _**nsxiv**_ really needs? Or are we trying to make up for the inadequacies of K1ss package manager?

I don't think anyone was trying to do the latter. I was trying to solve the problem not having to specify every optional dependency I don't use in PC manner. I've said before simply disabling all dependencies by default works for me. I believe others expressed autodetection is nice thing to have and don't believe any particular package manager was the motivating factor.


>If the package manger of a source based distro cannot keep track of optional dependencies then (it's a complete failure and) that package manager should be fixed, not the upstream build system.

May we please stay on topic and not attack irrelevant software. I feel like comments like this are just going to start a flame war. I've believe I've stated many times about my motivations. I gave the (former) kiss package of sxiv as an example of how autodetection would save marginal amount of effort for those interested. That wasn't (at least to me) a deciding factor but just an example of how it could be useful and probably applies to other source based distros.

TAAPArthur commented:
> we broke `make clean install` due to this (on our current release),

@N-R-K Also how is it broken? `make clean install` works for me on 27.1 (24e70a99e3). Can you share the output?

eylles commented:
> > we broke `make clean install` due to this (on our current release),
> 
> @N-R-K Also how is it broken? `make clean install` works for me on 27.1 ([24e70a9](https://github.com/nsxiv/nsxiv/commit/24e70a99e38ba71ac26e8476c5f2d4352c6ff548)). Can you share the output or describe the problem? It builds for me and the binary starts. It builds with `-DHAVE_LIBGIF=0 -DHAVE_LIBEXIF=0` is that the problem?
> 
> If so that's due the the config.mk not being included when clean is specified (to avoid it being created if it didn't exist) and the defaults being minimal dependencies.

https://github.com/nsxiv/nsxiv/issues/70#issuecomment-922286267

it broke for the user.

@DusanLesan can you provide the output you had when building? that would greatly help fix this.

N-R-K commented:
> can you provide the output you had when building? that would greatly help fix this.

No need, we already know why:
> If so that's due the the config.mk not being included when clean is specified 

And yes, since this deviates from the original sxiv (and other suckless software) I see this as a problem.

Also in the current state of this PR `make clean` will generate the auto detect only for it to be deleted and we dont have any standard way to avoid that.

One more thing to consider is that now we have limited changing depedancies to just CLI commands where before anyone could edit the Makefile without any knowledge about make.


> May we please stay on topic and not attack irrelevant software.

I put that into parenthesis specifically because that was not the main point, main point is we're trying to make up for a problem which is not ours (as in nsxiv)



> probably applies to other source based distros.

Is that an assumption or can you provide any real examples of this? We currently have it packaged for 3 distros. On arch it did nothing, on void it did nothing, on gentoo it made things more difficult (although not by much) as the auto detection needed to be sed-ed out.

TAAPArthur commented:
>One more thing to consider is that now we have limited changing depedancies to just CLI commands where before anyone could edit the Makefile without any knowledge about make.

You can still edit the Makefile. You can hard code whatever defaults there or in config.mk or pass them in the cli. Only thing that changed is default behavior

>Also in the current state of this PR make clean will generate the auto detect only for it to be deleted and we dont have any standard way to avoid that.

We have a standard way. We don't currently have a POSIX way. 

> Is that an assumption or can you provide any real examples of this? We currently have it packaged for 3 distros. On arch it did nothing, on void it did nothing, on gentoo it made things more difficult (although not by much) as the auto detection needed to be sed-ed out.

Did nothing in the since that you are compiling with all optional dependencies already installed right? In that case yes you get no benefit. That has nothing to do with the packing system which is why I don't understand why it keeps being brought up. Instead of focusing on the package manager you could imagine a user building from source who just wants the program to build and doesn't want to either install all the depeds or find all the flags they need to disable. 

Also instead of using sed to remove it, `touch config.mk` would have the same effect as disabling the autodetection given it does already exists.

N-R-K commented:
> You can still edit the Makefile

I believe you didn't read the last part.
> where before anyone could edit the Makefile without any knowledge about make.

Granted, figuring out `?=` shouldn't be difficult, but it's more difficult than it used to be.

> That has nothing to do with the packing system which is why I don't understand why it keeps being brought up.

From what I can tell the entire reason you want this is because [k1ss doesn't have any easy way to manipulate depedancies](https://github.com/kiss-community/kiss/pull/31) and you want to avoid forking. Which is understandable, but that's a k1ss problem not nsxiv's build problem.

Also do we have a POSIX way to deal with `make clean` generating auto detection? Not a big issue, but it is an issue that _did not_ exist before.

TAAPArthur commented:
> where before anyone could edit the Makefile without any knowledge about make.

You need knowledge of Make to edit makefiles. Saying that the `?=` is worse/less readable than `lib_exif_0/lib_exif_1` is subjective. Also my response was to 
>now we have limited changing depedancies to just CLI commands

Which to me seemed to claim that the only way to change dependencies was via the cli

>From what I can tell the entire reason you want this is because k1ss doesn't have any easy way to manipulate depedancies and you want to avoid forking. Which is understandable, but that's a k1ss problem not nsxiv's build problem.

No. That statement is not true. I've said many times it isn't true. I'm trying to clearly state for the nth time that my goal was just to build with minimal dependencies and that making all the optional ones non-default would be fine. Like I meant what I said earlier
> As far as I'm concerned, making nsxiv build with minimal dependencies by default would be sufficient. Thought auto-detecting would be more user friendly but I personally don't use any of the optional libs so not completely against removing the auto detection.

Like we could move the discussion to what the default behavior should be to build all optional deps or none of them. Or we could keep trying to assert package manager superiority. I really I don't believe I put up an objection to removing it. I submitted the PR because I thought it would be useful to people. If there's a reason I keep responding it is mainly because I don't agree with most/all of the points you are making. Especially the opinionated was about readability/user friendliness. 

>Also do we have a POSIX way to deal with make clean generating auto detection? Not a big issue, but it is an issue that did not exist before.

I think you could manage something with having sub-makefiles but don't think it is worth it

TAAPArthur commented:
But honestly searching for my past PRs on other reasons and then using it to claim my motivations when I'm directly stating something different is, too me, going too far.

N-R-K commented:
> > now we have limited changing depedancies to just CLI commands
> 
> Which to me seemed to claim that the only way to change dependencies was via the cli

why would you quote just the first part? I clearly stated
> where before anyone could edit the Makefile without any knowledge about make.

> But honestly searching for my past PRs

I didn't go through your past PRs. Yesterday (or the day before) you said something about PRs being rejected for philosophical reason. I was curious and simply looked at the very first PR on your profile that wasn't on nsxiv.

N-R-K commented:
Also in the original PR https://github.com/nsxiv/nsxiv/pull/19#issue-993768356
you stated

> Users could not get the extra optional features by just having those extra dependencies installed without having to modify the default build file.

which is a deficiency of k1ss pkg management. But now you're saying that has nothing to do with it??

I'm not a mind reader, so if someone states two contradictory things I can't really trust either and will have to just go with my guy.

N-R-K commented:
And with all that being said, currently my main problem is `make clean` running auto gen. If we can fix it POSIXly without breaking `make clean install` then I'll still approve the auto detection.

I just wanted to know clearly on why this is needed in nsxiv to begin with.

N-R-K commented:
Looked into sub-makefiles, and yeah it's not worth it. Also due to https://github.com/nsxiv/nsxiv/issues/85 I had to look into what a lot of the distros are doing with `sxiv` and I couldn't find any where they would benefit from auto-detection. Given these facts, autodetection is dropped once again.

If someone can point out one more distro that would benefit from auto-detection then I will be open to adding it back. Also we're now using `?=` on `HAVE_LIBX` so one can export those env var and disable an optional lib without touching the Makefile.

- - -

I don't think it's an option to disable the _current_ optional deps by default since that means breaking from what `sxiv` does and dealing with issues of users not having gif playback by default.

As for what the policy will be for _future_ optional dependencies, I'm okay having them disabled by default. Although, that does create an inconsistency where some optional deps are enabled by default but others aren't.

TAAPArthur commented:
> And yeah, this discussion went out of hand and I'm certainly to blame for that. But as I said, I felt like I needed to know the proper motivation behind the changes so that I can decide more clearly on if it's actually beneficial to nsxiv as a project or just a personal thing

@N-R-K I appreciate this comment and let me return the good faith. 



>> Users could not get the extra optional features by just having those extra dependencies installed without having to modify the default build file.

> which is a deficiency of k1ss pkg management. But now you're saying that has nothing to do with it??

So the reason I find the above comment unproductive/offensive is that what you call a "deficiency" is an intended feature and one I appreciate. But we aren't here to talk about our personal preferences so I didn't want to get into why it is helpful. I think the point it is that this aspect of dependency handling isn't a defect from the maintainers/users/my view. We could still care or not care about this usecase, but I feel like calling it a defect and a means to support your argument is unproductive, opinionated and derails the conversation.

Now you wanted a use case for autodetection. I don't believe I said anything contradictory I was stating one use case for the feature would make it easier for kiss users. The same is true for the AUR. For either package manger you could specify all the optional dependencies and this feature appears useless to those who install directly from the upstream repository. I think the disconnect here is that some users like me try to minimize dependencies. So if I install nsxiv form the AUR I wouldn't necessarily want libgif or libexif (my opinion) which are currently listed as mandatory. And it isn't hard to just edit the build file/PKGBUILD and remove the dependencies but then I have to know they are optional and then know the flags to remove them. If I'm just looking which aren't clear just form the PKGBUILD. 

I believe the main difference between AUR/kiss is the culture with the latter generally (at least that is the ideal) going for minimal dependencies in the official repos while the AUR leaves it more to the maintainer/unspecified.
With autodetection you could (arguably) just mark the optional dependencies as optional and or remove them completely. This way the maintainer doesn't have to force their views on all users by default. Now the maintainer can still decide what the default dependencies are to force them to be installed if they want to. I'd still argue the autodetect has benefit since it makes it easier for users to add/remove dependencies since they just have to touch a list rather than remember special vars. 

I think we all agree this isn't a critical feature. I believe it provides a minor convenience of letting users not edit build scripts for simple usecases. But again my goal is at the larger issue of how optional dependencies are enabled/disabled by default and this also extends making more existing depends optional.

So that's my long rant about some motivation that is completely aside from the point of the discussion. Right now we are deciding between 
1. being fully POSIX and removing autodetection,
2. keeping auto-detection as-is and being mainly POSIX and supported by every distro any of us can think of, 
3. making auto-detection optional in such a way that the Makefile is POSIX by default. 
4. dropping autodetection completely 

And at least my opinion would be affected by how optional dependencies are handled. I think we should handle all optional dependencies the same way and I find having all of them always-enabled by default bad.

If a tanglible exaple helps, we just added webp support. Now if we remove autodetection and I don't have libwebp installed, should by build break or not break?

N-R-K commented:
Thanks for you post, that cleared a couple things up. I think we can now move forward leaving the package manager discussion behind and focus on the actual point.

- - -

## So let's say we drop autodetection.

* Do we have all optional deps disabled by default? IMHO, this should be off the tables as we don't want user's coming from sxiv and losing gif playback, which is a pretty common format.

* Do we disable all the future optional deps by default? Personally I'm _OK_ with this. But as I've said it does create some inconsistency problem where some deps are enabled by default and others aren't.

## Lets say we keep auto-detection

* In order to stay POSIX we'll now have the auto-detection running whenever `make clean` is run. This annoys me a bit, but since this is nothing too fatal I might let it slide.

* Do we advise editing `autodetect.mk` ? This creates a bit of a fracture, where now certain things are edited in `autodetect.mk` and certain things in Makefile.

* Do we have auto-detection by default? Or should it instead be a separate rule `make auto-detect` ? If we go for the second one then I think that serves your usecase (?) but at the same time lets users opt-out of autodetect and edit the Makefile directly to their liking.

- - -

***Since not having auto-detection keeps things much simpler, I would prefer moving in that direction.*** Although I haven't mentioned it up until now, I also feel like autodetection isn't quite in line with the `suckless` mindset. This isn't exactly a technical argument, or an argument at all, but rather something worth considering given that sxiv is one of the more well-known suckless-style software.

TAAPArthur commented:
Wait so I'd like to challange one of the key assumptions we are making that `-include` isn't POSIX. From `man 1p make`

> Makefile  command  lines can have one or more of the following prefixes: a <hyphen-minus> ('-'), an at-sign ('@'), or a <plus-sign> ('+').  These shall modify the way in which make processes the command.
       -     If the command prefix contains a <hyphen-minus>, or the -i option is present, or the special target .IGNORE has either the current target as a  prerequisite or has no prerequisites, any error found while executing the command shall be ignored.

And having a makefile like
```
.POSIX:
-include config.mk
```
behaviors as expected when config.mk doesn't exist. So I claim that behavior is standard and not a problem.

So right now we have code to autodetect depends an generate a config file. In the same way we could generate a config file that enables/disables all the deps.

And I believe we are subjecting ourselves to the following constraints.
1. We shouldn't auto create the config file since it would run on `make clean` and we don't want to use ifs
2. Either building with all or none should be the default. Auto shouldn't be the default because it involves auto creating the config file.
3. (optional) The other 2 options should be targets that will create this config file. We could do without the convenience but it seems nice to have.

And if you are with me so far, should we just put it to a vote about what the default behavior is and if autodetection should be kept in the Makefile or converted into a patch?

I'm sure you could guess, but my vote would be keeping autodetection but building minimally by default. Maybe another question should be if `make clean` should remove the generated config file or not. I'd lean towards treating it like a config.h and not removing it and encouraging people to edit it.

N-R-K commented:
If `-include` is POSIX then we can use it. And btw `.POSIX` doesn't disable non-posix extensions, it simply means that if there's a local extension which conflicts with the posix spec, treat it posixly instead of locally.

- - -

@TAAPArthur Btw, as of now (in this PR) we're using `?=` for all the `HAVE_LIBX`, so if it's packaged with minimal deps, can't you just export those env vars and enable whatever lib you need without having to mess with the pkg/build system? If so then that should solve your usecase, no?

TAAPArthur commented:
>And btw .POSIX doesn't disable non-posix extensions,

Reading is hard.

> Btw, as of now (in this PR) we're using `?=` for all the `HAVE_LIBX`, so if it's packaged with minimal deps, can't you just export those env vars and enable whatever

That is basically the same as disabling them manually. Yeah you can do it. Whatever is done will break with any new additional dependency which is less than ideal. Like ideally would be able to build minimally and have the build not break when knew optional depends are added.

N-R-K commented:
> Whatever is done will break with any new additional dependency which is less than ideal. Like ideally would be able to build minimally and have the build not break when knew optional depends are added.

I see your point, but adding new deps will not be a regular occurrence, in fact it's going to be a very rare occurrence. Which is why I feel like this use-case may not be worth the trouble.

TAAPArthur commented:
Perhaps but by the same logic changing the default to be use minimal will also be a rare occurrence. Also I feel like a similar statement was made about condig.def.h. 

Outside of changing the default from sxiv, is there a downside to a minimal build by default? The config file is already incompatible and would imagine for most people adding extra flags to make would be easier than converting their config. I also imagine the majority of users are somewhat comfortable applying patches so specifying makeflags shouldn't be a concern. If it is just a matter of who is in majority, should we just vote?

N-R-K commented:
> Outside of changing the default from sxiv, is there a downside to a minimal build by default?

I would assume that an image viewer is capable of playing `gif`s, even if we ignore the backwards compatibility issue with `sxiv`, this is still a sensible default. Also if the status bar is to be made an optional feature, do we disable that by default too? I don't really see that being a good move.

> If it is just a matter of who is in majority, should we just vote?

Not exactly, my main concern with having all opt deps disabled by default is that we break backwards compatibility with `sxiv` and I don't believe it's sensible for an image viewer to not be able to play `gif`s by default.

But if we want to do a vote on this, then I am fine with it.

TAAPArthur commented:
>Also if the status bar is to be made an optional feature, do we disable that by default too?

One of the reasons I sent that PR out now. Figured it would be relevant to the discussion

> I don't believe it's sensible for an image viewer to not be able to play gifs by default.

And here I was thinking people didn't even use gifs regularly.

>But if we want to do a vote on this, then I am fine with it.

Not trying to brush off the issue or just kill the discussion, but think we just have conflicting opinions/principles. Whatever the non-default option is can just be a make target like config.h (like auto-generated and never cleaned) so it could be run once and forget. So once it is resolved we can both get the behavior we want.

There's also the option of forging being perfectly POSIX or reworking the autodetection to avoid the issue. I had once proposed doing the auto detection in an optional configure script. Guess it doesn't solve the problem of what the default would be but could make it easy to change.

N-R-K commented:
> I had once proposed doing the auto detection in an optional configure script.

We decided to drop that due to this https://github.com/nsxiv/nsxiv/pull/19#issuecomment-917565639 (which I think was a valid criticism)

Let's say we enable everything by default (no auto-detection), is there a way that would allow users to easily build with all opt deps disabled. Is it feasible to have something like `make minimal` which would do that?

N-R-K commented:
Very hacky, but take a look @TAAPArthur. Does this solve your use-case?

TAAPArthur commented:
> We decided to drop that due to this [#19 (comment)](https://github.com/nsxiv/nsxiv/pull/19#issuecomment-917565639) (which I think was a valid criticism)

I don't think anything brought up there was fatal.  I think the arguments about "well-known conventions" are often abused when not citing a specific source. I don't think this applies to this PR since POSIX is at least a standard, but I've seem plenty of projects where configure is optional.


> Very hacky, but take a look @TAAPArthur. Does this solve your use-case?

I guess; I don't like the solution. I was toying around with the idea of make minimal/all but don't we reject changes that can be done with sed like we did with the whole `static const` PR.

Also I'd prefer to generate a config file that has the desired settings once than have to type `make minimal` every time. Having make minimal generate a minimal config that is auto-included could work. 

I'm also still in in opposition to the default being all optional dependencies always on by default. I believe on stances on this are mutually exclusive so don't think there is room for compromise without autodetection. And even with autodetect I'm not sure if we agree that the default (in case the config file is empty) should be enabled or disabled.

N-R-K commented:
> don't we reject changes that can be done with sed like we did with the whole `static const` PR.

That was changing the source code as well for a usecase that was very much personal.

> Also I'd prefer to generate a config file that has the desired settings once than have to type make minimal every time.

That can be done via a wrapper function in `.bashrc`.

As for default disabling opt deps, my stance remains the same, we can't disable what was by default enabled on `sxiv` but I'm fine if everyone else thinks disabling future opt deps is the right move.

And yes, it seems like you're right. It seems very unlikely that we can agree on a what the default will be, so we may just have to settle this on a vote.

N-R-K commented:
Since the only problem with this current approach can easily be solved via a wrapper function, and is merely a convenience issue, I guess we're moving forwards with `make minimal` and dropping auto-detection.

So now the question is:

## Should nsxiv enable or disable all optional deps by default?

***NOTE: The statusbar might become an optional feature in the future.*** See: https://github.com/nsxiv/nsxiv/pull/95

* React :+1:  for enabling all optional deps by default; `make minimal` would disable all opt deps.
* React :-1: for disabling all optional deps by default; `make maximum` would enable all opt deps.

@eylles @bakkeby @GRFreire @kdkasad @XPhyro @TAAPArthur

TAAPArthur commented:
> That can be done via a wrapper function in `.bashrc`.
> Since the only problem with this current approach can easily be solved via a wrapper function,

A wrapper is a pretty heavy handed workaround that is still more inconvenient. I guess you could override make and have it know what you want by the current wd, but still pretty convoluted. Also don't find it very clean and you yourself have called it hacky. I agree it is doable, I just don't think anyone would do it vs modifying the Makefile itself which isn't unreasonable.



> Once we can decide on what the default should be (and update the docs accordingly), I think this PR is will be ready to be rebased and merged.

I get the feeling that enabling all deps by default is winning. Unfortunate, but moving on to talk about the implementation.

To summarize what has already been presented and/or stuff of the top of my head
1. Temp files; We create a temp makefile, replace the defaults, call make and then delete the temp file. Think this is hacky and should be avoided. At the time of writing, this is what is implemented in the PR.
2. Generate a config.mk (or whatever we want to call it) file in the same way we do config.h. As in create it from a template/autogenerated iff it doesn't already exit. This way users can set it up once and just pretend their way is the default. Also works well for people who want something in between all off/all on or just want the exact set of depends regardless of what upstream does. It also clearly spells out all libs that are abled to be toggled without the other noise in the Makefile.
3. Have a var like `HAVE_LIB_DEFAULT ?= 1` that controls the default value of optional libs. Users could just change this value to toggle all on all off and then specify whatever libs they want to explicitly include/exclude. Not mutually exclusive from (2).
4. Like (2) but also recreate and remove it with clean. This way `make minimal` `make clean` `make` will build with the defaults instead of minimal. Would also need to warn users to not edit the config file directly

Feel free to add more. Having an optional config.mk file seems the most favorable to extension. 

>I guess we're moving forwards with make minimal and dropping auto-detection.

Also @N-R-K I had just meant to withdraw objection to removing autodetection not consenting to removing it. I don't think there is harm in it being optional. 

P.S
> Note: Users are also encouraged to vote, as this will affect them the most

Aren't we all affected in the same given we build from source?

eylles commented:
Regardless number of commits we cannot take this as a personal pet project, we always have to have users in mind, a balance between what users want and need and hwta the developer wants is the space where a project thrives, projects with toxic devs out there are a dime the dozen.

TAAPArthur commented:
> Regardless number of commits we cannot take this as a personal pet project, we always have to have users in mind, a balance between what users want and need and hwta the developer wants is the space where a project thrives, projects with toxic devs out there are a dime the dozen.

Was this statement made in response to something I said? Or was it just a general commit?

eylles commented:
in general really, i do think every decision that affects the codebase in any significant way a consideration should be made of how benefitial it is to the userbase not just for ourselves.

N-R-K commented:
> Having an optional config.mk file seems the most favorable to extension.

Imo that should be avoided as it just fractures the build time configuration. If we do want to go for some config.mk it shouldn't just be libraries, we should move all Makefile vars onto it similar to other suckless projects.

- - -

> Have a var like HAVE_LIB_DEFAULT ?= 1 that controls the default value of optional libs.

Can this be done inside the Makefile POSIX-ly ? If so then I'm on board.

TAAPArthur commented:
> Imo that should be avoided as it just fractures the build time configuration. If we do want to go for some config.mk it shouldn't just be libraries, we should move all Makefile vars onto it, similar to other suckless projects. See: https://git.suckless.org/dwm/file/config.mk.html

I don't understand what you mean by "fractures". It optionally allows users to use a config file. I don't see the problem with giving people extra options that can opt-in to. Feel like a similar argument can be made for Xresource and some cli args along with the same counterargument that they don't hurt people who don't care about them. Sticking `-include config.mk` at the top of the makefile seems harmless and simple enough along with having some targets auto generate it. The problem I see with including it as part of the repo is that it prevents us from overriding it via make and it being a tad harder for users to bring their own config.mk like they do for config.h. Also don't see a point in hard coding it if we can generate it. 

I don't object to including more things than just libs. Users could add whatever they wanted into it as stated. We could create it with all the default args that are in the Makefile to better show what users can override I guess. 

>Can this be done inside the Makefile POSIX-ly ? If so then I'm on board.

Wouldn't it be something like:
```
HAVE_LIB_DEFAULT ?= 1
HAVE_LIB_WEBP ?= $(HAVE_LIB_DEFAULT)
HAVE_LIB_GIF ?= $(HAVE_LIB_DEFAULT)
...
```
And this is just my personal opinion here, but with this, I don't think we'd need a make minimal. I'd be content with just dropping `-include.mk` at the top and if someone wanted they could just `echo "HAVE_LIB_DEFAULT ?= 1" > config.mk`.

N-R-K commented:
> Wouldn't it be something like:
> 
> ```
> HAVE_LIB_DEFAULT ?= 1
> HAVE_LIB_WEBP ?= $(HAVE_LIB_DEFAULT)
> HAVE_LIB_GIF ?= $(HAVE_LIB_DEFAULT)
> ...
> ```

I was thinking more in terms of something like `NSXIV_MINIMAL_BUILD` and have that disable inotify as well. But I guess this works too.

But if we allow a var to disable all deps, then I don't see much reason to also add `include whatever.mk` since that env var can just be exported.

TAAPArthur commented:
> But if we allow a var to disable all deps, then I don't see much reason to also add `include whatever.mk` since that env var can just be exported.

By that logic shouldn't every config.mk not exist since the any var could also be exported? Exporting everything either risks polluting the user's env namespace and passing vars on the command line is more key strokes when building. Nothing fatal, but at least that's why I see benefit in this extra line. You just drop the settings in and don't have to keep thinking about them if you don't want to but are free to specify them every time too. Also helps if you want something inbetween all off or all on like if you want just the set that sxiv needed or something. It isn't specific to dependencies. You could also set `PREFIX` or `AUTORELOAD` with the same change.

What's the downside of adding the extra line?

N-R-K commented:
> What's the downside of adding the extra line?

It's including a file that won't exist. Why not just go full suckless mode with a `config.mk` and put all configurable vars in there?

TAAPArthur commented:
> > What's the downside of adding the extra line?
> 
> It's including a file that won't exist. Why not just go full suckless mode with a `config.mk` and put all configurable vars in there?

Including a file that _may_ not exists. It is a harmless if it doesn't exist and has merit if it does. Also opens the downs for a more sane `make minimal`/`make auto`. You haven't listed a downside. 

As for why this may be more convient than including a config.mk in the repo... same reason config.def.h is more convenient than having config.h directly in the repo. So people can just drop their own settings in.

N-R-K commented:
> Including a file that _may_ not exists. It is a harmless if it doesn't exist and has merit if it does. Also opens the downs for a more sane `make minimal`/`make auto`. You haven't listed a downside.

Potential confusion as this file doesn't exist nor will it be generated by anything. I guess we'll just have to deal with it via adding a comment.

- - -

So are we just going with `HAVE_LIB_DEFAULT` which only disables optional libs or do we try to disable `inotify` as well?

0mp commented on Makefile:
```diff
@@ -11,27 +6,25 @@ PREFIX ?= /usr/local
+lib_gif_0 =
+lib_gif_1 = -lgif
+LDLIBS = -lImlib2 -lX11 -lXft -lfontconfig \
+  $(lib_exif_$(HAVE_LIBEXIF)) $(lib_gif_$(HAVE_LIBGIF))
```
Would be cool to have a way of specifying additional  libraries. For example, on FreeBSD it's necessary to append `-linotify` to the this list.

N-R-K replied:
We can use `+=` instead.

Anyways is there any other distro you can think of where this will be useful? I don't think we want to make changes if it's just *one* distro specific problem. If there are more than we're OK making changes.

0mp replied:
I'm not sure about the other but OpenBSD would benefit from this interface as well: http://cvsweb.openbsd.org/cgi-bin/cvsweb/ports/graphics/sxiv/Makefile?rev=1.17&content-type=text/x-cvsweb-markup
-

0mp replied:
Homebrew is also modifying LDLIBS: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/sxiv.rb

0mp replied:
Excellent, thanks.

N-R-K commented:
Still not fully sold on having `-include custom.mk` ... but anyways what should the comment be @TAAPArthur ?

0mp commented:
> Still not fully sold on having `-include custom.mk` ... but anyways what should the comment be @TAAPArthur ?

You may avoid `-include` by instructing the users that if customization are necessary they can create a file called `makefile` where they 

1. put their custom variables
2. add `include Makefile`

Since `makefile` has a precedence over `Makefile`, running `make` is going to read `makefile` instead of `Makefile`.

N-R-K commented:
> > Still not fully sold on having `-include custom.mk` ... but anyways what should the comment be @TAAPArthur ?
> 
> You may avoid `-include` by instructing the users that if customization are necessary they can create a file called `makefile` where they
> 
>     1. put their custom variables
> 
>     2. add `include Makefile`
> 
> 
> Since `makefile` has a precedence over `Makefile`, running `make` is going to read `makefile` instead of `Makefile`.

That works nicely, thanks. I guess we're not doing any ghost includes in the Makefile then.

N-R-K commented:
Squashed, rebased and ready for review.

If there's any remaining concerns, please raise them now. After this, the build system should not receive any major changes. We can't keep changing our build system and making life harder for package maintainers.

TAAPArthur commented:
> You may avoid `-include` by instructing the users that if customization are necessary they can create a file called `makefile` where they

This is a great solution. I completely forgot make checks other things besides `Makefile` Wounder how much of that is historical vs for this specific usecase. That satisfies me and is much better than the ghost include

TAAPArthur commented on Makefile:
```diff
@@ -12,36 +7,37 @@ MANPREFIX = $(PREFIX)/share/man
 DOCPREFIX = $(PREFIX)/share/doc/nsxiv
```
`=` -> `?=` for MANPREFIX and DOCPREFIX

N-R-K replied:
I think it's natural that man and doc goes under the same PREFIX. And one can use `make MANPREFIX=..` to change if needed.

I'm fine keeping this as is, but okay changing it too.

0mp replied:
Actually, unless examples are directly linked to some documentation documents, it would be nice to be able to install examples separately from documents. On FreeBSD for example we install documentation at $(PREFIX)/share/doc/foobar and examples at $(PREFIX)/share/examples/foobar. From what I know, other systems have similar standards.

N-R-K replied:
@0mp Our "document" here is basically just the examples (key-hander and image-info). So you can do `make DOCPREFIX="$PREFIX/share/examples/nsxiv"` to install them under the desired directory.

N-R-K replied:
That would create `$PREFIX/share/examples/nsxiv/examples/{key-handler,image-info}` , is that ok or did you want `$PREFIX/share/examples/nsxiv/{key-handler,image-info}` ?

N-R-K replied:
@0mp Since we aren't really installing any "docs" but only examples, just changed it to `EGPREFIX` instead.

TAAPArthur commented on Makefile:
```diff
@@ -12,36 +7,37 @@ MANPREFIX = $(PREFIX)/share/man
 
 # autoreload backend: inotify/nop
-AUTORELOAD = inotify
+AUTORELOAD ?= inotify
```
Should we take this opportunity to change `AUTORELOAD` to `HAVE_INOTIFY` for consistency?

N-R-K replied:
How would we implement it? I was thinking about using preprocessor ifdefs but decided to drop that idea because that would mean if someone changes AUTORELOAD it wont trigger a rebuild.

TAAPArthur replied:
I don't think changes to makeflags should trigger a rebuild just like enabling/disabling libs won't trigger a rebuild. I think ifdefs are fine and to compliment that maybe we should just merge `autoreload_inofity` and `autoreload_nop` instead of having reverse ifdefs in each

N-R-K replied:
I guess we can do what we're doing with the opt libs. Very hacky, but since `ifeq` isn't posix, not much else I can think of.

N-R-K replied:
Fixed in https://github.com/nsxiv/nsxiv/pull/71/commits/12c0881de16dec2eb8d7474825bbf70c9d021f63
